wasm.go

```diff
@@ -1,66 +1,171 @@
+	"github.com/tetratelabs/wazero/internal/wasm/interpreter"
+	"github.com/tetratelabs/wazero/internal/wasm/jit"
 	"github.com/tetratelabs/wazero/internal/wasm/text"
+	"github.com/tetratelabs/wazero/wasm"
 )
 
-// DecodeModule parses the configured source into a Module. This function returns when the source is exhausted or
-// an error occurs. The result can be initialized for use via InstantiateModule.
-//
-// Here's a description of the return values:
-// * result is the module parsed or nil on error
-// * err is a FormatError invoking the parser, dangling block comments or unexpected characters.
-// See binary.DecodeModule and text.DecodeModule
-type DecodeModule func(source []byte) (result *Module, err error)
-
-var DecodeModuleBinary DecodeModule = func(source []byte) (*Module, error) {
-	return decodeModule(binary.DecodeModule, source)
+type Engine struct {
+	e internalwasm.Engine
 }
 
-var DecodeModuleText DecodeModule = func(source []byte) (*Module, error) {
-	return decodeModule(text.DecodeModule, source)
+func NewEngineInterpreter() *Engine {
+	return &Engine{e: interpreter.NewEngine()}
 }
 
-func decodeModule(decoder internalwasm.DecodeModule, source []byte) (*Module, error) {
-	m, err := decoder(source)
-	if err != nil {
-		return nil, err
+func NewEngineJIT() *Engine { // TODO: compiler?
+	return &Engine{e: jit.NewEngine()}
+}
+
+// StoreConfig allows customization of a Store via NewStoreWithConfig
+type StoreConfig struct {
+	// Context is the default context used to initialize the module. Defaults to context.Background.
+	//
+	// Notes:
+	// * If the Module defines a start function, this is used to invoke it.
+	// * This is the outer-most ancestor of wasm.ModuleContext Context() during wasm.HostFunction invocations.
+	// * This is the default context of wasm.Function when callers pass nil.
+	//
+	// See https://www.w3.org/TR/wasm-core-1/#start-function%E2%91%A0
+	Context context.Context
+	// Engine defaults to NewEngineInterpreter
+	Engine *Engine
+}
+
+func NewStore() wasm.Store {
+	return internalwasm.NewStore(context.Background(), interpreter.NewEngine())
+}
+
+// NewStoreWithConfig returns a store with the given configuration.
+func NewStoreWithConfig(config *StoreConfig) wasm.Store {
+	ctx := config.Context
+	if ctx == nil {
+		ctx = context.Background()
 	}
-	var name string
-	if m.NameSection != nil {
-		name = m.NameSection.ModuleName
+	engine := config.Engine
+	if engine == nil {
+		engine = NewEngineInterpreter()
 	}
-	return &Module{wasm: m, name: name}, nil
+	return internalwasm.NewStore(ctx, engine.e)
 }
 
-// EncodeModule encodes the given module into a byte slice depending on the format of the implementation.
-// See binary.EncodeModule
-type EncodeModule func(m *Module) []byte
-
-var EncodeModuleBinary EncodeModule = func(m *Module) []byte {
-	return encodeModule(binary.EncodeModule, m)
+// ModuleConfig defines the WebAssembly 1.0 (MVP) module to instantiate.
+type ModuleConfig struct {
```

Comment:
PS there's no chance of misinterpreting text for binary format as the magic is not valid text. However, if later we find a user who really wants to not detect, we can add a flag to force the encoding of the source. I'm intentionally not doing that as each knob should pull its weight and I really don't think this would be worth describing.